paragraph/bridging/hlo/hlo_converter.cc

```diff
@@ -0,0 +1,300 @@
+/* Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "paragraph/bridging/hlo/hlo_converter.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "paragraph/bridging/hlo/hlo_compute_cost_analysis.h"
+#include "tensorflow/compiler/xla/service/hlo_module.h"
+#include "tensorflow/compiler/xla/util.h"
+
+xla::StatusOr<paragraph::GraphProto> HloModuleToGraphProto(
+    const xla::HloModule* module,
+    const ComputeCostAnalysis* cost_analysis,
+    bool reset_ids);
+xla::StatusOr<paragraph::SubroutineProto> HloComputationToSubroutineProto(
+    const xla::HloComputation* computation,
+    const ComputeCostAnalysis* cost_analysis,
+    int64_t new_id);
+
+const std::string HloToParagraphOpcodeString(
+    xla::HloOpcode hlo_opcode) {
+  if (hlo_opcode == xla::HloOpcode::kWhile) {
+    return "while";
+  } else if (hlo_opcode == xla::HloOpcode::kCall) {
+    return "call";
+  } else if (hlo_opcode == xla::HloOpcode::kConditional) {
+    return "conditional";
+  } else if (hlo_opcode == xla::HloOpcode::kAllGather) {
+    return "all-gather";
+  } else if (hlo_opcode == xla::HloOpcode::kAllReduce) {
+    return "all-reduce";
+  } else if (hlo_opcode == xla::HloOpcode::kAllToAll) {
+    return "all-to-all";
+  } else if (hlo_opcode == xla::HloOpcode::kSend) {
+    return "send-start";
+  } else if (hlo_opcode == xla::HloOpcode::kSendDone) {
+    return "send-done";
+  } else if (hlo_opcode == xla::HloOpcode::kRecv) {
+    return "recv-start";
+  } else if (hlo_opcode == xla::HloOpcode::kRecvDone) {
+    return "recv-done";
+  } else if (hlo_opcode == xla::HloOpcode::kInfeed) {
+    return "infeed";
+  } else if (hlo_opcode == xla::HloOpcode::kOutfeed) {
+    return "outfeed";
+  } else {
+    return "delay";
+  }
+  return "delay";
+}
+
+xla::HloCostAnalysis::Properties GetPropertiesForInstruction(
+    const int64_t id,
```

Comment:
There is no reason to use const for types that are passed by value. There are a couple of these in here.

Reply:
Fixed